Update Windows Symbol Visibility

ooeygui opened the pull request:
This change adds a dedicated page for symbol visibility on Windows.
The expanded text attempts to clear up confusing around Symbol Visibility on Windows,
In order to improve the porting experience for users.

Signed-off-by: Lou Amadio <lamadio@microsoft.com>

ooeygui committed:
Add Dedicated page for Windows Symbol Visibility

This change adds a dedicated page for symbol visibility on Windows.
The expanded text attempts to clear up confusing around Symbol Visibility on Windows,
In order to improve the porting experience for users.

Signed-off-by: Lou Amadio <lamadio@microsoft.com>

clalancette reviewed (changes requested):
I'd rather expand the "Windows Tips and Tricks" document rather than add a new Tutorial here.  Also, I think we can significantly shorten it by showing just the parts of `visibility_control.h` that pertain to Windows, and then have a link out to a full example.

Finally, please do one sentence per-line, which matches the style of the rest of the document.

traversaro commented on source/Tutorials/Miscellaneous/Windows-Symbol-Visibility.rst:
```diff
@@ -0,0 +1,159 @@
+   #endif
+   #define MY_LIB_PUBLIC_TYPE
+   #endif
+   #endif  // MY_LIB__VISIBILITY_CONTROL_HPP_
```
If I am not wrong, this version of the header would also decorate with `__declspec(dllexport)` and `__declspec(dllimport)` also static libraries (as opposed to the example at the end of https://gcc.gnu.org/wiki/Visibility). Do you think this is ok? If building static libraries with this visibility headers, perhaps the support for static libraries can be listed as an advantage of `WINDOWS_EXPORT_ALL_SYMBOLS`?

ooeygui replied:
Thanks for the feedback, I'll take a look :-)

ooeygui replied:
WINDOWS_EXPORT_ALL_SYMBOLS doesn't export static class members - which are common with Qt UI. Here's a good writeup of that - https://www.kitware.com/create-dlls-on-windows-without-declspec-using-new-cmake-export-all-feature.

traversaro replied:
Sorry, I probably was not clear, I actually was referring to static **libraries** (i.e. compiling with `BUILD_SHARED_LIBS` set to `OFF`), not static class members. To clarify, it is perfect fine for me if the tutorial does not cover the case of static libraries, I just think it would be helpful to clarify that.

ooeygui replied:
Sorry for the delay - I see what you mean. I agree - I'll work on clarifying wording.

ooeygui replied:
I started looking into adding BUILD_SHARED_LIB off support, and it was getting a bit icky. Would you be ok if I add a task to come back and work on that later?

traversaro replied:
Sure! I wanted just to make sure that the fact that `BUILD_SHARED_LIBS` set to `OFF` is not supported is clear.

traversaro commented on source/Tutorials/Miscellaneous/Windows-Symbol-Visibility.rst:
```diff
@@ -0,0 +1,159 @@
+      #define MY_LIB_EXPORT __declspec(dllexport)
+      #define MY_LIB_IMPORT __declspec(dllimport)
+   #endif
+   #ifdef MY_LIB_BUILDING_LIBRARY
```
Perhaps it should be documented that `MY_LIB_BUILDING_LIBRARY` should be defined in someway when the library is compiled, for example via [`DEFINE_SYMBOL`](https://cmake.org/cmake/help/v3.23/prop_tgt/DEFINE_SYMBOL.html), that by default is `MY_LIB_EXPORTS`  if the library is called `MY_LIB` .

traversaro replied:
See also https://github.com/ros2/ros2_documentation/pull/2820#discussion_r917120088 .

ooeygui commented:
Addressed PR feedback. (although, the title is no longer accurate - I've moved the dedicated page to the tips and tricks section as requested)

Thanks all!

clalancette commented on source/How-To-Guides/Ament-CMake-Documentation.rst:
```diff
@@ -255,6 +255,8 @@ To use it for a package called ``my_library``:
 
     target_compile_definitions(my_library PRIVATE "MY_LIBRARY_BUILDING_LIBRARY")
 
+For more details, see :doc:`Windows Symbol Visibility in the Windows Tips and Tricks document</The-ROS2-Project/Contributing/Windows-Tips-and-Tricks>`.
```
```suggestion
For more details, see :doc:`Windows Symbol Visibility in the Windows Tips and Tricks document <../The-ROS2-Project/Contributing/Windows-Tips-and-Tricks>`.
```

(it is a little weird that we use relative doc links like this, but this is consistent with the rest of the documentation)

clalancette commented on source/The-ROS2-Project/Contributing/Windows-Tips-and-Tricks.rst:
```diff
@@ -27,23 +27,95 @@ This has two practical consequences for ROS 2:
-In the ROS 2 core code, packages often have a header called ``visibility_macros.h`` that use macros to define the proper ``dllexport``/``dllimport`` statements.
-With these macros in place, any symbol in header files that needs to be public needs to be decorated with the ``PUBLIC`` version of the macro.
-The CMakeLists.txt of the package should also have a stanza like:
+The Microsoft Visual C++ Compiler (MSVC) exposes symbols from a Dynamic Link Library (DLL) only if they are explicitly exported. The clang and gcc compilers have an option to do the same, but it is off by default. As a result, when a library previously built on Linux is built on Windows, other libraries may be unable to resolve the external symbols.
```
```suggestion
The Microsoft Visual C++ Compiler (MSVC) exposes symbols from a Dynamic Link Library (DLL) only if they are explicitly exported.
As a result, when a library previously built on Linux is built on Windows, other libraries may be unable to resolve the external symbols.
```

(essentially, one sentence per-line.  The same throughout the rest of this document)

ooeygui replied:
I think I got them all.

clalancette commented on source/The-ROS2-Project/Contributing/Windows-Tips-and-Tricks.rst:
```diff
@@ -27,23 +27,95 @@ This has two practical consequences for ROS 2:
+^^^^^^^^^^^^^^^^^^^^^^^^^^
+The purpose of Visibility Control Headers headers is to define a macro for each library which correctly declares symbols as dllimport or dllexport. This is decided based on whether the library is being consumed or being built itself. The logic in the macro also takes the compiler into account and includes logic to select the appropriate syntax.
+The following link includes step by step instructions for adding explicit symbol visibility to a library “yielding the highest quality code with the greatest reductions in binary size, load times and link times”: [The GCC visibility documentation](https://gcc.gnu.org/wiki/Visibility).
+A header named ``visibility_control.hpp`` can be placed in the ``includes`` folder for each library as shown in the example below.
```
```suggestion
A header named ``visibility_control.h`` can be placed in the ``include`` folder for each library as shown in the example below.
```

(that is, it can be used for both C and C++, and also we tend to use 'include', not 'includes')

clalancette commented on source/The-ROS2-Project/Contributing/Windows-Tips-and-Tricks.rst:
```diff
@@ -27,23 +27,95 @@ This has two practical consequences for ROS 2:
 
 .. code-block:: cmake
 
-  target_compile_definitions(${PROJECT_NAME}
-    PRIVATE "XXX_BUILDING_LIBRARY")
```
Wait, we still need this bit in "Visibility Control Headers", right?  That is, we need to tell CMake to set XXX_BUILDING_LIBRARY, otherwise the exports won't be done correct.

clalancette commented on source/The-ROS2-Project/Contributing/Windows-Tips-and-Tricks.rst:
```diff
@@ -27,23 +27,95 @@ This has two practical consequences for ROS 2:
+
+In these cases dllimprort/dllexport must be applied explicitly. This can be done using generate_export_header as described in the following article: `Create dlls on Windows without declspec() using new CMake export all feature <https://blog.kitware.com/create-dlls-on-windows-without-declspec-using-new-cmake-export-all-feature/>`__.
 
-Finally, it is important that the header file that exports the symbols be included into at least one of the ``.cpp`` files in the package so that the macros will get expanded and placed into the resulting binary.
-Otherwise the symbols will still not be callable.
```
Personally, I think this is valuable advice to leave in.  I have hit this problem several times, and I have also seen contributors struggle with this as well.

ooeygui replied:
Good catch! I'll work it back in

clalancette commented on source/Tutorials/Intermediate/Writing-an-Action-Server-Client/Cpp.rst:
```diff
@@ -71,7 +71,7 @@ Go into the action workspace you created in the :doc:`previous tutorial <../Crea
 
 In order to make the package compile and work on Windows, we need to add in some "visibility control".
-For details on why this is needed, see `here <https://docs.microsoft.com/en-us/cpp/cpp/dllexport-dllimport>`_.
+For more details, see :doc:`Windows Symbol Visibility in the Windows Tips and Tricks document</The-ROS2-Project/Contributing/Windows-Tips-and-Tricks>`.
```
```suggestion
For more details, see :doc:`Windows Symbol Visibility in the Windows Tips and Tricks document <../The-ROS2-Project/Contributing/Windows-Tips-and-Tricks>`.
```

ooeygui committed:
Update source/How-To-Guides/Ament-CMake-Documentation.rst

Co-authored-by: Chris Lalancette <clalancette@gmail.com>

ooeygui committed:
Update source/The-ROS2-Project/Contributing/Windows-Tips-and-Tricks.rst

Co-authored-by: Chris Lalancette <clalancette@gmail.com>

ooeygui committed:
Update source/Tutorials/Intermediate/Writing-an-Action-Server-Client/Cpp.rst

Co-authored-by: Chris Lalancette <clalancette@gmail.com>

ooeygui committed:
Address PR Feedback

Signed-off-by: Lou Amadio <lamadio@microsoft.com>

ooeygui committed:
Fix Linting break

clalancette committed:
Additional cleanup.

Signed-off-by: Chris Lalancette <clalancette@openrobotics.org>

clalancette reviewed (approved):
I just went through and did some additional cleanup here.  With that, this looks good to me so I'll approve.  Thanks for the contribution!
